jetty-http2/http2-server/src/main/java/org/eclipse/jetty/http2/server/HttpTransportOverHTTP2.java

```diff
@@ -352,117 +353,259 @@ public void abort(Throwable failure)
 
     private class TransportCallback implements Callback
     {
-        private State state = State.IDLE;
-        private Callback callback;
-        private Throwable failure;
-        private boolean commit;
+        private State _state = State.IDLE;
+        private Callback _callback;
+        private boolean _commit;
+        private Throwable _failure;
 
-        public boolean start(Callback callback, boolean commit)
+        private void reset()
         {
-            State state;
+            assert Thread.holdsLock(this);
+            _state = State.IDLE;
+            _callback = null;
+            _commit = false;
+            _failure = null;
+        }
+
+        private void send(Callback callback, boolean commit, Consumer<Callback> consumer)
+        {
+            Throwable failure = preSend(callback, commit);
+            if (failure == null)
+            {
+                consumer.accept(this);
+                postSend();
+            }
+            else
+            {
+                callback.failed(failure);
+            }
+        }
+
+        private Throwable preSend(Callback callback, boolean commit)
+        {
+            synchronized (this)
+            {
+                switch (_state)
+                {
+                    case IDLE:
+                    {
+                        _state = State.PRE_SEND;
+                        _callback = callback;
+                        _commit = commit;
+                        return null;
+                    }
+                    default:
+                    {
+                        Throwable failure = _failure;
+                        if (failure == null)
+                            failure = new IllegalStateException("Invalid transport state: " + _state);
+                        return failure;
+                    }
+                }
+            }
+        }
+
+        private void postSend()
+        {
+            Callback callback;
+            boolean commit;
             Throwable failure;
             synchronized (this)
             {
-                state = this.state;
-                failure = this.failure;
-                if (state == State.IDLE)
+                switch (_state)
                 {
-                    this.state = State.WRITING;
-                    this.callback = callback;
-                    this.commit = commit;
-                    return true;
+                    case PRE_SEND:
+                    {
+                        _state = State.POST_SEND;
+                        return;
+                    }
+                    case SUCCEED:
+                    {
+                        callback = _callback;
```

Comment:
Some comments on how we could get to these states between preSend and postSend would be good.